File: Leetcode/966/966.go
func g(s *string) string {
	var sb strings.Builder
	sb.Grow(len(*s))
	for i := 0; i < len(*s); i++ {
		ch := unicode.ToLower(rune((*s)[i]))
		if ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u' {
			sb.WriteRune('a')
		} else {
			sb.WriteRune(ch)
		}
	}
	return sb.String()
}

func spellchecker(wordlist []string, queries []string) []string {
	unset := make(map[string]struct{})
	empty := struct {}{}
	s := ""
	for i := 0; i < len(wordlist); i++ {
		s = wordlist[i]
		unset[s] = empty
	}
	unmap := make(map[string]string)
	for i := 0; i < len(wordlist); i++ {
		s = strings.ToLower(wordlist[i])
		_, ok := unmap[s]
		if !ok {
			unmap[s] = wordlist[i]
		}
	}
	unmap2 := make(map[string]string)
	for i := 0; i < len(wordlist); i++ {
		s = g(&wordlist[i])
		_, ok := unmap2[s]
		if !ok {
			unmap2[s] = wordlist[i]
		}
	}
	ans := make([]string, len(queries))
	for i := 0; i < len(queries); i++ {
		_, ok := unset[queries[i]]
		if ok {
			ans[i] = queries[i]
			continue
		}
		s = strings.ToLower(queries[i])
		val, ok := unmap[s]
		if ok {
			ans[i] = val
			continue
		}
		s = g(&queries[i])
		val, ok = unmap2[s]
		if ok {
			ans[i] = val
		}
	}
	return ans
}

File: Leetcode/966/966.cpp
class Solution {
public:
    string f(string s) {
        for (auto &ch : s) {
            ch = tolower(ch);
        }
        return s;
    }
    string g(string s) {
        for (auto &ch : s) {
            ch = tolower(ch);
            if (ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u') {
                ch = 'a';
            }
        }
        return s;
    }
    vector<string> spellchecker(vector<string>& wordlist, vector<string>& queries) {
        int n = (int)queries.size();
        int m = (int)wordlist.size();
        unordered_set<string> unset(wordlist.begin(),wordlist.end());
        unordered_map<string, string> unmap;
        string s;
        for (int i = 0; i < m; ++i) {
            s = f(wordlist[i]);
            if (unmap.find(s) == unmap.end()) {
                unmap[s] = wordlist[i];
            }
        }
        unordered_map<string, string> unmap2;
        for (int i = 0; i < m; ++i) {
            s = g(wordlist[i]);
            if (unmap2.find(s) == unmap2.end()) {
                unmap2[s] = wordlist[i];
            }
        }
        vector<string> ans(n);
        for (int i = 0; i < n; ++i) {
            if (unset.find(queries[i]) != unset.end()) {
                ans[i] = queries[i];
                continue;
            }
            s = f(queries[i]);
            if (unmap.find(s) != unmap.end()) {
                ans[i] = unmap[s];
                continue;
            }
            s = g(queries[i]);
            if (unmap2.find(s) != unmap2.end()) {
                ans[i] = unmap2[s];
            }
        }
        return ans;
    }
};